### PR TITLE
REF: separate indexer utilities from indexing.py

### PR DESCRIPTION
**Pull request by jbrockmendel**
A couple things going on here.

- Take simple, low-dependency, not-used-in-indexing.py functions out of indexing.py and put them in indexers.py.
  - A couple of runtime-defined functions in internals get moved here too.
  - Avoids some runtime imports (i.e. simplify dependency structure)
  - Fixes some incorrect private names
  - Adds typing
- Simplify some indexing.py methods
  - Avoid duplicated validation calls
  - Add typing
  - De-duplicate _get_slice_axis method
  - Move code outside of try/except blocks to clarify what the failure modes are.

Medium-term goals in indexing.py
  - Isolate internals access
  - refactor _setitem_with_indexer (over 300 lines)

### CODE REVIEW

File: pandas/core/indexers.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+from pandas.core.dtypes.common import is_list_like
+from pandas.core.dtypes.generic import ABCIndexClass, ABCSeries
+
+
+# -----------------------------------------------------------
+# Indexer Identification
+
+def is_list_like_indexer(key) -> bool:
+    # allow a list_like, but exclude NamedTuples which can be indexers
```
Comment (jreback) [inline pandas/core/indexers.py]: can u add as proper doc-strings

File: pandas/core/indexers.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    return is_list_like(key) and not (isinstance(key, tuple) and type(key) is not tuple)
+
+
+def is_scalar_indexer(indexer, arr_value) -> bool:
```
Comment (jreback) [inline pandas/core/indexers.py]: if possible to type pls do so (may be tricky)

I would type with Any explicity if it’s truly that
so we know this is done (meaning it has been typed)

possible to use reveal_type to figure things out here

**Review by jreback — CHANGES_REQUESTED** [review]
just some quick comments
will have a full look later

+1 on refactoring things

**Comment by jreback** [comment]
note as a post or pre-cursor to this I would form pandas/core/indexing/ and move indexing.py (and indexers.py) there as well

**Comment by jbrockmendel** [comment]
> note as a post or pre-cursor to this I would form pandas/core/indexing/ and move indexing.py (and indexers.py) there as well

Sure.  Obvious preference for post-cursor.

File: pandas/core/indexers.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexers.py]: ATM I've added types for everything I had figured out with sufficient certainty.  I'd rather leave it clearly-unfinished so it gets more attention than put Any if we can be more specific

File: pandas/core/indexers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexers.py]: sure

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -1329,20 +1308,16 @@ def _tuplify(self, loc):
...
         if not need_slice(slice_obj):
             return obj.copy(deep=False)
-        indexer = self._convert_slice_indexer(slice_obj, axis)
 
-        if isinstance(indexer, slice):
-            return self._slice(indexer, axis=axis, kind="iloc")
-        else:
-            return self.obj._take(indexer, axis=axis)
+        indexer = self._convert_slice_indexer(slice_obj, axis)
+        assert isinstance(indexer, slice), type(indexer)
```
Comment (jreback) [inline pandas/core/indexing.py]: prob better off to have this assert in _slice

File: pandas/core/indexing.py [hunk 4]
```diff
@@ -1450,11 +1425,11 @@ class _LocationIndexer(_NDFrameIndexer):
     def __getitem__(self, key):
         if type(key) is tuple:
```
Comment (jreback) [inline pandas/core/indexing.py]: should prob use isinstance here as this is an old idiom

**Review by jreback — CHANGES_REQUESTED** [review]
just a couple of doc-strings I think (in the first section), then can merge this

**Comment by jbrockmendel** [comment]
Added one more docstring, left un-docstringed is_scalar_indexer and is_empty_indexer because I'm not totally clear on what those functions are doing (possibly an extra indentation in is_scalar_indexer?  it is _almost_ identical to the last 5 lines of is_empty_indexer)

Added a note about InvalidIndexError being for geopandas.

File: pandas/core/indexing.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexing.py]: tried this, turns out this is intentional to get tuple but exclude NamedTuple.  Added a comment to that effect.  (green)

File: pandas/core/indexers.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexers.py]: can you fix this doc-sring in followup and type as much as possible here

File: pandas/core/indexers.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            indexer = tuple([indexer])
+        return any(isinstance(idx, np.ndarray) and len(idx) == 0 for idx in indexer)
+    return False
+
+
+# -----------------------------------------------------------
+# Indexer Validation
+
+
+def check_setitem_lengths(indexer, value, values) -> None:
```
Comment (jreback) [inline pandas/core/indexers.py]: type

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. in followups we need to move / consolidate these things; also a typing pass would be helpful.

**Pull request merged**